test/functional/test_framework/mininode.py

```diff
@@ -440,3 +454,127 @@ def network_thread_join(timeout=10):
     for thread in network_threads:
         thread.join(timeout)
         assert not thread.is_alive()
+
+class P2PStub(P2PInterface):
+    """A P2P Stub class.
+
+    Keeps a block and transaction store and responds correctly to getdata and getheaders requests."""
+
+    def __init__(self):
+        super().__init__()
+        self.reject_code_received = None
+        self.reject_reason_received = None
+        self.block_store = OrderedDict()
+        self.tx_store = {}
+        self.getdata_requests = []
+
+    def on_getdata(self, message):
+        """Check for the tx/block in our stores and if found, reply with an inv message."""
+        for inv in message.inv:
+            self.getdata_requests.append(inv.hash)
+            if inv.type == 1 or inv.type == 1 | MSG_WITNESS_FLAG:
+                if inv.hash in self.tx_store.keys():
+                    self.send_message(msg_tx(self.tx_store[inv.hash]))
+            elif inv.type == 2 or inv.type == 2 | MSG_WITNESS_FLAG:
+                if inv.hash in self.block_store.keys():
+                    self.send_message(msg_block(self.block_store[inv.hash]))
```

Comment:
Mind to simplify the nested ifs?

```py
elif inv.type & MSG_TX and inv.hash in self.block_store.keys():
    self.send...

Reply:
inv types are not bit flags in general, so `inv.type & MSG_TX` evaluates to true for MSG_FILTERED_BLOCK (and `inv.type & MSG_BLOCK` also evalutes to true for MSG_FILTERED_BLOCK).

Flattening the nested ifs would give:

```python
            if (inv.type == 1 or inv.type == 1 | MSG_WITNESS_FLAG) and inv.hash in self.tx_store.keys():
                self.send_message(msg_tx(self.tx_store[inv.hash]))
            elif (inv.type == 2 or inv.type == 2 | MSG_WITNESS_FLAG) and inv.hash in self.block_store.keys():
                self.send_message(msg_block(self.block_store[inv.hash]))
```

I'm not sure if that's any clearer due to the length of the if statements, but I'm happy to change it if you think it's an improvement.

Reply:
Nope, fine to leave as is.

Reply:
taken @ryanofsky's suggested change